Make sentence_transformers an optional dependency

theparthgupta opened the pull request:
## Summary
This PR makes `sentence_transformers` an optional dependency as requested in #636. The dependency is heavy and not needed for all use cases, so users can now choose whether to install it.

## Changes Made
- **pyproject.toml**: Moved `sentence_transformers` from required dependencies to optional dependencies under the `sentence-transformers` extra
- **python/cocoindex/features.py**: Added conditional import logic to only define `SentenceTransformerEmbed` when the dependency is available
- **python/cocoindex/__init__.py**: Updated to conditionally export `SentenceTransformerEmbed` only when available
- **examples/**: Added installation instructions to all example files that use `SentenceTransformerEmbed`

## Installation Options
Users can now install with:
```bash
# With sentence-transformers support
pip install 'cocoindex[sentence-transformers]'

# Or install the dependency separately
pip install sentence-transformers

Fixes #636

lemorage commented:
Hey, personally, I feel that for the extra name, `sentence-transformers` is a bit long and does not quite make sense. Following [PEP 621](https://peps.python.org/pep-0621/#dependencies-optional-dependencies), **Extras** collect specific subsets of optional dependencies based on **features**. Maybe we can change it to some descriptive names like `ml` or `embeddings`. This way, it is also quite extensible for future changes.

theparthgupta commented on examples/amazon_s3_embedding/main.py:
```diff
@@ -1,3 +1,13 @@
+Or install the dependency directly:
+    pip install sentence-transformers
+"""
+
```
Sure! I'll update the documentation and add sentence-transformers to the pyproject.toml files for the relevant examples. Give me a bit to make those changes.

theparthgupta commented:
I've updated all 7 examples that use `SentenceTransformerEmbed` with the new optional dependency approach.

I noticed that the main README.md and several docs files also reference `SentenceTransformerEmbed`. Should I update those files with the new installation instructions as well, or would you prefer to handle the documentation updates yourself?

Files that might need updates:
- README.md (main example uses SentenceTransformerEmbed)  
- docs/quickstart.md
- docs/basics.md
- docs/functions.md
- docs/custom_function.mdx

badmonster0 commented:
> Hey, personally, I feel that for the extra name, `sentence-transformers` is a bit long and does not quite make sense. Following [PEP 621](https://peps.python.org/pep-0621/#dependencies-optional-dependencies), **Extras** collect specific subsets of optional dependencies based on **features**. Maybe we can change it to some descriptive names like `ml` or `embeddings`. This way, it is also quite extensible for future changes.

Agree, `sentence-transformers` is indeed too long and probably too specific. +1 for using `embedding` as the feature name.

badmonster0 commented:
> I've updated all 7 examples that use `SentenceTransformerEmbed` with the new optional dependency approach.
> 
> I noticed that the main README.md and several docs files also reference `SentenceTransformerEmbed`. Should I update those files with the new installation instructions as well, or would you prefer to handle the documentation updates yourself?
> 
> Files that might need updates:
> 
> * README.md (main example uses SentenceTransformerEmbed)
> * docs/quickstart.md
> * docs/basics.md
> * docs/functions.md
> * docs/custom_function.mdx

It'll be great if you can help to update these docs in the current PR:

* docs/quickstart.md (in the instruction for `pip install cocoindex`, we need to add this new optional feature)
* docs/functions.md (we need to tell users this `SentenceTransformerEmbed` function needs this new optional feature)

Other docs may be just mentioning `SentenceTransformerEmbed`. Seems they don't need to be updated.

`README.md` usually doesn't need to be updated either - after the dependency is added into `pyproject.toml`, the dependency will be automatically installed in `pip install`.

Thanks!

theparthgupta commented on examples/amazon_s3_embedding/main.py:
```diff
@@ -1,3 +1,13 @@
+Or install the dependency directly:
+    pip install sentence-transformers
+"""
+
```
will remove these too

theparthgupta commented:
I've addressed all the feedback regarding documentation updates:
Changes made:
- ✅ **docs/quickstart.md**: Updated pip install command to `pip install -U 'cocoindex[sentence-transformers]'`
- ✅ **docs/functions.md**: Added optional dependency note for `SentenceTransformerEmbed` function with clear installation instructions
- ✅ **All 7 examples**: Removed manual installation instructions from README.md and main.py files as requested

theparthgupta commented on docs/docs/getting_started/quickstart.md:
```diff
@@ -1,294 +1,294 @@
-title: Quickstart
-description: Get started with CocoIndex in 10 minutes
----
+    ---
```
reverted this

theparthgupta commented:
> > > Hey, personally, I feel that for the extra name, `sentence-transformers` is a bit long and does not quite make sense. Following [PEP 621](https://peps.python.org/pep-0621/#dependencies-optional-dependencies), **Extras** collect specific subsets of optional dependencies based on **features**. Maybe we can change it to some descriptive names like `ml` or `embeddings`. This way, it is also quite extensible for future changes.
> > 
> > 
> > Agree, `sentence-transformers` is indeed too long and probably too specific. +1 for using `embedding` as the feature name.
> 
> @theparthgupta what do you think about this? Would you also change the feature name to `embedding`? Thanks!

Agreed on changing it! I'd lean toward embeddings (plural) since it typically handles multiple embeddings, but either embedding or embeddings works well. Much better than the current long name. If you agree i can work on this too

badmonster0 commented:
> embeddings

`embeddings` looks good. Thanks!

theparthgupta commented:
> > embeddings
> 
> `embeddings` looks good. Thanks!

Sounds good! I'll get started on the implementation.

theparthgupta commented:
I've updated the optional dependency group name from sentence-transformers to embeddings as discussed. Let me know if you'd like any other naming adjustments or additions

theparthgupta commented:
> Thanks for this change!

thanks for merging the PR!

 I noticed CI is failing on main after my PR was merged — looks like mypy can't find sentence_transformers or numpy. I think it's due to the optional import. Should I open a quick fix to update mypy.ini or CI deps?

lemorage commented:
@theparthgupta The CI failure is largely due to the missing `numpy` dep, after making `sentence_transformers` optional. I've made the fix, and feel free to check out #676.
